### PR TITLE
refactor: Integrate with opendal for s3

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

Hi, this is a quick demo for https://github.com/mozilla/sccache/issues/1404. This PR intends to show what opendal's API looks like and how opendal will work with sccache. And give the whole community a chance to try [opendal](https://github.com/datafuselabs/opendal).

This PR only changes the s3 implementation. In face, we can implement `Cache` for `opendal::Operator` so that we don't need to implement `Cache` for different storage backends anymore.

Any comments are welcome! Thank you so much for your attention and participation.

Fix https://github.com/mozilla/sccache/issues/1386
Fix https://github.com/mozilla/sccache/issues/1404
Fix https://github.com/mozilla/sccache/issues/633

### CODE REVIEW

**Comment by Xuanwo** [comment]
Seems our crates.io index need a update 😲

**Comment by messense** [comment]
sccache currently requires MSRV 1.58.0, so opendal also need to support Rust 1.58.0 unless the maintainers are willing to update to a new MSRV policy.

**Comment by Xuanwo** [comment]
> sccache currently requires MSRV 1.58.0, so opendal also need to support Rust 1.58.0 unless the maintainers are willing to update to a new MSRV policy.

Thanks for explanation!

**Comment by sylvestre** [comment]
What is the MSRV for opendal?

**Comment by sylvestre** [comment]
Needs rebasing now, thanks :)

File: src/cache/s3.rs [hunk 1]
```diff
@@ -40,25 +33,52 @@ impl S3Cache {
...
+        builder.root(key_prefix);
+        if no_credentials {
+            builder.disable_credential_loader();
+        }
+        if let Some(endpoint) = endpoint {
+            let endpoint_uri: http::Uri = endpoint
+                .try_into()
+                .map_err(|err| anyhow!("input endpoint {endpoint} is invalid: {:?}", err))?;
+            let mut parts = endpoint_uri.into_parts();
+            match use_ssl {
```
Comment (sylvestre) [inline src/cache/s3.rs:50]: could you please keep this as in function (last the previous version)

File: src/cache/s3.rs [hunk 2]
```diff
@@ -203,24 +123,8 @@ mod test {
...
-        assert_eq!(
-            normalize_key("prefix/", "0123456789abcdef0123456789abcdef"),
-            "prefix/0/1/2/0123456789abcdef0123456789abcdef"
+            normalize_key("0123456789abcdef0123456789abcdef"),
+            "0/1/2/0123456789abcdef0123456789abcdef"
         );
     }
-
-    #[test]
-    fn test_endpoint_resolver() {
```
Comment (sylvestre) [inline src/cache/s3.rs:216]: please keep this test

File: src/cache/s3.rs (hunk 1 above)
Comment (Xuanwo) [reply src/cache/s3.rs]: Added back.

File: src/cache/s3.rs (hunk 2 above)
Comment (Xuanwo) [reply src/cache/s3.rs:216]: Added back.

**Comment by drahnr** [comment]
There are a few risks associated with `opendal`, it..

* .. depends on `tokio`, so we are signing up for that, not a big deal  *I think*, we already use tokio heavily and I don't see that changing
* .. has a lot backends we don't need, i.e. `rocksdb` - these are configurable via feature flags in `main` in `opendal` - and that'd be one condition to not inflate the compile time cost for zero benefit to `sccache`
* .. future proofness is an open topic, that wasn't exactly addressed
* .. the stability guarantees are also not entirely clear to me on an API level

On the other side, we gain quite a bit:

* Simplified addition of more backends
* The sampled code and documentation looks actually better than what we have right now
*  Shared load of maintainence web service API changes

My bottom line is a `yay` given the manpower that's available to maintain `sccache`

**Comment by sylvestre** [comment]
Great summary, exactly how i feel...

maybe we could do a new release of sccache with the current code (and a few PR)
and take this into a new release

**Comment by Xuanwo** [comment]
> * .. has a lot backends we don't need, i.e. `rocksdb` - these are configurable via feature flags in `main` in `opendal` - and that'd be one condition to not 
>           inflate
>            the compile time cost for zero benefit to `sccache`

Most backends like `rocksdb` are only included while the corresponding feature enabled like: https://opendal.databend.rs/opendal/#services-1

They will not add the compile time cost for `sccache` unless enabling them.

Services that only need HTTP client will be enabled by default (which includes services like s3, gcs, oss) because they will not introduce extra heavy dependences.

If this is a big concern to `sccache`, I'm willing to add new feature flags for them too.

---

Answers to open questions:

> * Future proof-ness, how to assure continued maintainence effort?

Yes. OpenDAL is the core dependence of [databend](https://github.com/datafuselabs/databend/) and we will continue to maintain it. And we are kept enlarge our community to get more maintainers.

So far, we have two maintainers: @Xuanwo and @PsiACE.

> * What are the stability guarantees on an API level, are they fully semver compliant?

Yes, we adopt the semver compliant, all changes will be recorded at https://github.com/datafuselabs/opendal/blob/main/CHANGELOG.md

NOTE: OpenDAL is on the track on `v0.X.Y` now, we could introduce breaking changes between `v0.X` version.

We are working hard to make a `v1.0` release.

> * MSRV handling?

OpenDAL's MSRV is `1.60` which is covered by our CI services: https://github.com/datafuselabs/opendal/blob/main/.github/workflows/ci.yml#L42

We will not change the MSRV before `v1.0` versions.

---

> maybe we could do a new release of sccache with the current code (and a few PR)

LGTM.

**Comment by drahnr** [comment]
@Xuanwo thank you for the prompt response, it addressed my concerns, for me it's good to go, right after:

@sylvestre we should cut a release before merging this PR as you suggested, just to get some experience/burning time with it

**Comment by drahnr** [comment]
@sylvestre I think it's time to merge

**Comment by sylvestre** [comment]
@Xuanwo you are going to update the various backend next, right ?

**Comment by Xuanwo** [comment]
> @Xuanwo you are going to update the various backend next, right ?

Yes, I plan to migrate `azure`, `gcs` first and than adding new backends support.

**Comment by sylvestre** [comment]
Github is complaining about conflicts

**Pull request merged**